metatomic-torch/src/outputs.cpp

```diff
@@ -0,0 +1,548 @@
+#include <torch/script.h>
+#include <algorithm>
+#include <cassert>
+#include <metatensor/torch.hpp>
+#include <ranges>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "./internal/utils.hpp"
+#include "metatomic/torch/exports.h"
+#include "metatomic/torch/model.hpp"
+#include "metatomic/torch/system.hpp"
+
+namespace metatomic_torch {
+
+const std::array energy_bases = {"energy", "energy_ensemble", "energy_uncertainty"};
+const std::array energy_gradients = {"strain", "positions"};
+
+torch::Dtype to_torch_dtype(const caffe2::TypeMeta& meta) {
+    return c10::typeMetaToScalarType(meta);
+}
+
+std::vector<std::string> split(const std::string& s, char delimiter) {
+    std::vector<std::string> result;
+    size_t start = 0;
+    size_t end = 0;
+    while ((end = s.find(delimiter, start)) != std::string::npos) {
+        result.push_back(s.substr(start, end - start));  // safe
+        start = end + 1;
+    }
+    result.push_back(s.substr(start));
+    return result;
+}
+
+std::string join_names(const std::vector<std::string>& names) {
+    std::ostringstream oss;
+    oss << "[";
+    for (size_t i = 0; i < names.size(); i++) {
+        oss << "'" << names[i] << "'";
+        if (i + 1 < names.size()) {
+            oss << ", ";
+        }
+    }
+    oss << "]";
+    return oss.str();
+}
+
+void _validate_single_block(const std::string& name,
+                            const metatensor_torch::TensorMap& value) {
+    // Ensure the TensorMap has a single block with the expected key
+
+    const auto valid_label = metatensor_torch::LabelsHolder::create({"_"}, {{0}});
+    const auto incoming_label = value->keys();
+    if (*valid_label != *incoming_label) {
+        C10_THROW_ERROR(ValueError, "invalid keys for \'" + name +
+                                        "\' output: expected " + valid_label->str() +
+                                        ", get: " + incoming_label->str());
+    }
+}
+
+void _validate_atomic_samples(
+    const std::string& name,
+    const metatensor_torch::TensorMap& value,
+    const std::vector<System>& systems,
+    const ModelOutput& request,
+    const torch::optional<metatensor_torch::Labels>& selected_atoms) {
+    // Validates the sample labels in the output against the expected structure
+
+    const torch::Device& device = value->device();
+    const metatensor_torch::TensorBlock& block = value->block_by_id(value, 0);
```

Comment:
`block_by_id` is a static method in C++, and should be called as `TensorMap::block_by_id(value, 0)`

Reply:
(they are static because they need to take `torch::intrusive_ptr<TensorMapHolder>` and not TensorMapHolder as `this`).